File: src/uri/P1883_Escape_From_Ayutthaya.py
# author : Teerapat Phokhonwong
# Onlinejudge: URI Online Judge
# Categories:
# Problem:
# Link:
# Timelimit: 1 sec
# Status: TLE
# Submission:
# Runtime: 
# Solution:
# Note:

N = M = 0
MAP = []
start = None
end = None
fire = None

#top -> bot -> right -> left
nY = [-1, 1, 0, 0]
nX = [0, 0, -1, 1]

class Point:
    x = y = t = 0
        
    def __init__(self,y, x, t=0):
        self.x = x
        self.y = y
        self.t = t
    
    def equals(self,p):
        return self.x == p.x and self.y == p.y 

    
def BFS(Q):
    while(len(Q) > 0):
        p = Q.pop(0)
        if p == None: 
            continue
    
        pX = p.x
        pY = p.y
        t = p.t
        MAP[pY][pX] = 1;
        dead = (abs(fire.x - pX) + abs(fire.y - pY))
        if dead == t : 
            continue
        
        if p.equals(end):
            return True

        for i in range(0, 4):
            pX = p.x + nX[i]
            pY = p.y + nY[i]
            if pX > 0 and pX < M and pY > 0 and pY < N :
                if MAP[pY][pX] == 0 :
                    nextPoint = Point(pY,pX,t+1)
                    Q.append(nextPoint)

    return False
    
    
#Main 
t = int(input())
while(t > 0):
    t = t-1
    N, M = map(int, input().split())
    MAP = []
    start = None
    end = None
    fire = None
    Q = []
    for y in range(0, N):
        MAP.append(list(input()))
        for x in range(len(MAP[y])):
            if MAP[y][x] == '.':
                MAP[y][x] = 0
            elif MAP[y][x] == '#':
                MAP[y][x] = 1
            elif MAP[y][x] == 'S':
                MAP[y][x] = 1
                start = Point(y,x)
            elif MAP[y][x] == 'E' :
                MAP[y][x] = 0
                end = Point(y,x);
            elif MAP[y][x] == 'F' :
                MAP[y][x] = 1
                fire = Point(y,x)
            else:
                MAP[y][j] = 0
    
    # for y in range(0, N):
    #     print(MAP[y])

    escape = (abs(start.x - start.x) + abs(start.y - start.y))
    dead = (abs(fire.x - end.x) + abs(fire.y - end.y))
    if escape >= dead :
        print("N")
        continue
    
    Q.append(start)
    answer = BFS(Q)
    # for y in range(0, N):
    #     print(MAP[y])
        
    if answer :
        print("Y")
    else:
        print("N")
#End Main 

File: cpp/beecrowd/P1883_Escape_From_Ayutthaya.cpp
/**
 * @author Teerapat Phokhonwong
 * @Onlinejudge: URI Online Judge
 * @Categories: AD-HOC
 * @Problem: 1883 - Zoeiro Keyboard
 * @Link: https://www.urionlinejudge.com.br/judge/en/problems/view/1883
 * @Timelimit: 1 sec
 * @LANGUAGE: C++17
 * @Status: Accepted
 * @Submission: 3/21/18, 2:13:14 AM
 * @Runtime: 0.412s
 * @Solution: BFS
 * @Note:
 */

#include <iostream>
#include <queue>

using namespace std;

const int MX = 1005;
const int BURNED = -1;
const int VISITED = 1;
const char WALL = '#';

char board[MX][MX];
int fire[MX][MX], n, m;
bool visit[MX][MX];

struct point {
    int x, y, t;

    point() {}

    point(int tx, int ty) : x(tx), y(ty) {}

    point(int tx, int ty, int tt) : x(tx), y(ty), t(tt) {}

};

int mx[] = {1, 0, -1, 0};
int my[] = {0, 1, 0, -1};

void fireSpread(vector<point> &f) {
    queue<point> Q;
    for (int j = 0; j < f.size(); ++j) {
        Q.push(f[j]);
        fire[f[j].x][f[j].y] = 1;
    }

    while (!Q.empty()) {
        point tm = Q.front();
        Q.pop();
        int t = tm.t + 1;
        for (int i = 0; i < 4; i++) {
            int tx = tm.x + mx[i];
            int ty = tm.y + my[i];
            if (tx >= 0 && tx < n
                && ty >= 0 && ty < m
                && board[tx][ty] != WALL
                && fire[tx][ty] == BURNED) {
                fire[tx][ty] = t;
                Q.push(point(tx, ty, t));
            }
        }
    }
}

bool escape(point &s, point &e) {
    queue<point> Q;
    Q.push(s);
    visit[s.x][s.y] = VISITED;
    while (!Q.empty()) {
        point tm = Q.front();
        Q.pop();
        if (tm.x == e.x && tm.y == e.y) {
            return true;
        }
        int t = tm.t + 1;
        for (int i = 0; i < 4; i++) {
            int tx = tm.x + mx[i];
            int ty = tm.y + my[i];
            if (tx >= 0 && tx < n
                && ty >= 0 && ty < m
                && board[tx][ty] != WALL
                && t < fire[tx][ty]
                && visit[tx][ty] == 0) {
                visit[tx][ty] = VISITED;
                Q.push(point(tx, ty, t));
            }
        }
    }
    return false;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        cin >> n >> m;
        vector<point> vf;
        point st, en;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                cin >> board[i][j];
                switch (board[i][j]) {
                    case 'F' :
                        vf.push_back(point(i, j, 1));
                        break;
                    case 'S':
                        st = point(i, j, 1);
                        break;
                    case 'E':
                        en = point(i, j);
                        break;
                }
                fire[i][j] = BURNED;
                visit[i][j] = 0;
            }
        }
        fireSpread(vf);
        cout << (escape(st, en) ? "Y" : "N") << endl;
    }
    return 0;
}
